### PR TITLE
feat(bulk-load): bulk load ingestion part6 - meta handle bulk_load_response during ingestion

### PR DESCRIPTION
**Pull request by hycdong**
The whole bulk load ingestion process is like:
1. meta server set app and all partitions bulk load status as ingesting #486 
2. meta send two requests to primary 
    - send ingestion request to primary #486 
    - send bulk load request to query ingestion status #457
3. primary handle two requests above:
    - primary consider ingestion request as a client write and replica group execute 2pc process to do actual rocksdb ingestion operation #490 pegasus pr [#546](https://github.com/XiaoMi/pegasus/pull/546)
    - primary handle bulk load request and broadcast it to secondary #496 #460
4. secondary handle two cases:
    - ingestion files through 2pc #490 pegasus pr [#546](https://github.com/XiaoMi/pegasus/pull/546)
    - report ingestion status to primary #496 
5. primary report group ingestion status to meta server #496 
6. meta handle two response
    - handle ingestion_response #493 
    - **handle group ingestion status and do bulk load status transformation**

This pull request is about meta handle ingestion status reported by replica server, part of step 6 above.
- meta handle ingestion status function `handle_app_ingestion`
    - check if `is_group_ingestion_finished` and replica's `ingest_status` are set, if not set, meta should not update partition ingestion_status in this round
    - if any replica's ingestion status is `IS_FAILED`, we consider it as an unrecoverable error, bulk load will set as failed
    - if all replicas of the group ingest succeed, meta will set this partition status from ingesting to succeed
- bulk load status transformation
    - if all partitions status turn from ingesting to succeed, meta will set app status from ingestion to succeed in `update_partition_status_on_remote_storage_reply`

### CODE REVIEW

File: src/dist/replication/meta_server/meta_bulk_load_service.cpp [hunk 1]
```diff
@@ -514,8 +514,58 @@ void bulk_load_service::handle_app_downloading(const bulk_load_response &respons
...
-    // TODO(heyuchen): TBD
-    // called by `on_partition_bulk_load_reply` when app status is ingesting
+    const std::string &app_name = response.app_name;
+    const gpid &pid = response.pid;
+
+    if (!response.__isset.is_group_ingestion_finished) {
+        dwarn_f("recevie bulk load response from node({}) app({}) partition({}), "
+                "primary_status({}), but is_group_ingestion_finished is not set",
+                primary_addr.to_string(),
+                response.app_name,
```
Comment (levy5307) [inline src/dist/replication/meta_server/meta_bulk_load_service.cpp:524]: `app_name`

File: src/dist/replication/meta_server/meta_bulk_load_service.cpp [hunk 2]
```diff
@@ -514,8 +514,58 @@ void bulk_load_service::handle_app_downloading(const bulk_load_response &respons
...
+        return;
+    }
+
+    for (const auto &kv : response.group_bulk_load_state) {
+        const auto &bulk_load_states = kv.second;
+        if (!bulk_load_states.__isset.ingest_status) {
+            dwarn_f("recevie bulk load response from node({}) app({}) partition({}), "
+                    "primary_status({}), but node({}) ingestion_status is not set",
+                    primary_addr.to_string(),
+                    response.app_name,
```
Comment (levy5307) [inline src/dist/replication/meta_server/meta_bulk_load_service.cpp:536]: `app_name`

File: src/dist/replication/lib/bulk_load/replica_bulk_loader.cpp [hunk 3]
```diff
@@ -565,9 +568,7 @@ void replica_bulk_loader::report_group_ingestion_status(/*out*/ bulk_load_respon
         is_group_ingestion_finish =
             is_group_ingestion_finish && (ingest_status == ingestion_status::IS_SUCCEED);
```
Comment (acelyc111) [inline src/dist/replication/lib/bulk_load/replica_bulk_loader.cpp:569]:
```suggestion
        is_group_ingestion_finish &&= (ingest_status == ingestion_status::IS_SUCCEED);
```

File: src/dist/replication/meta_server/meta_bulk_load_service.cpp [hunk 4]
```diff
@@ -514,8 +514,58 @@ void bulk_load_service::handle_app_downloading(const bulk_load_response &respons
 void bulk_load_service::handle_app_ingestion(const bulk_load_response &response,
                                              const rpc_address &primary_addr)
 {
-    // TODO(heyuchen): TBD
-    // called by `on_partition_bulk_load_reply` when app status is ingesting
+    const std::string &app_name = response.app_name;
+    const gpid &pid = response.pid;
+
+    if (!response.__isset.is_group_ingestion_finished) {
+        dwarn_f("recevie bulk load response from node({}) app({}) partition({}), "
```
Comment (acelyc111) [inline src/dist/replication/meta_server/meta_bulk_load_service.cpp:521]:
```suggestion
        dwarn_f("receive bulk load response from node({}) app({}) partition({}), "
```

File: src/dist/replication/meta_server/meta_bulk_load_service.cpp [hunk 5]
```diff
@@ -514,8 +514,58 @@ void bulk_load_service::handle_app_downloading(const bulk_load_response &respons
...
+                    primary_addr.to_string(),
+                    response.app_name,
+                    pid,
+                    dsn::enum_to_string(response.primary_bulk_load_status),
+                    kv.first.to_string());
+            return;
+        }
+
+        if (bulk_load_states.ingest_status == ingestion_status::IS_FAILED) {
+            derror_f("app({}) partition({}) node({}) ingestion failed",
```
Comment (acelyc111) [inline src/dist/replication/meta_server/meta_bulk_load_service.cpp:544]: Keep the same order as L533?

File: src/dist/replication/meta_server/meta_bulk_load_service.cpp [hunk 6]
```diff
@@ -514,8 +514,58 @@ void bulk_load_service::handle_app_downloading(const bulk_load_response &respons
...
+                     app_name,
+                     pid,
+                     kv.first.to_string());
+            handle_bulk_load_failed(pid.get_app_id());
+            return;
+        }
+    }
+
+    ddebug_f("recevie bulk load response from node({}) app({}) partition({}), primary_status({}), "
```
Comment (acelyc111) [inline src/dist/replication/meta_server/meta_bulk_load_service.cpp:553]:
```suggestion
    ddebug_f("receive bulk load response from node({}) app({}) partition({}), primary_status({}), "
```

File: src/dist/replication/meta_server/meta_bulk_load_service.cpp (hunk 5 above)
Comment (hycdong) [reply src/dist/replication/meta_server/meta_bulk_load_service.cpp]: This log format is like L474-L479, shows which node ingestion failed.

**Pull request merged**